src/ai/backend/manager/repositories/container_registry/db_source/db_source.py

```diff
@@ -0,0 +1,115 @@
+
+
+class ContainerRegistryDBSource:
+    """Database source for container registry-related operations."""
+
+    _db: ExtendedAsyncSAEngine
+
+    def __init__(self, db: ExtendedAsyncSAEngine) -> None:
+        self._db = db
+
+    async def fetch_by_registry_and_project(
+        self,
+        registry_name: str,
+        project: Optional[str] = None,
+    ) -> ContainerRegistryData:
+        """Fetch container registry data by registry name and optional project."""
+        async with self._db.begin_readonly_session() as session:
+            stmt = sa.select(ContainerRegistryRow).where(
+                ContainerRegistryRow.registry_name == registry_name,
+            )
+            if project:
+                stmt = stmt.where(ContainerRegistryRow.project == project)
+
+            row: Optional[ContainerRegistryRow] = await session.scalar(stmt)
+            if row is None:
+                raise ContainerRegistryNotFound()
+            return row.to_dataclass()
+
+    async def fetch_by_registry_name(self, registry_name: str) -> list[ContainerRegistryData]:
+        """Fetch all container registries with the given registry name."""
+        async with self._db.begin_readonly_session() as session:
+            stmt = sa.select(ContainerRegistryRow).where(
+                ContainerRegistryRow.registry_name == registry_name
+            )
+            result = await session.execute(stmt)
+            rows: list[ContainerRegistryRow] = result.scalars().all()
+            return [row.to_dataclass() for row in rows]
+
+    async def fetch_all(self) -> list[ContainerRegistryData]:
+        """Fetch all container registries."""
+        async with self._db.begin_readonly_session() as session:
+            stmt = sa.select(ContainerRegistryRow)
+            result = await session.execute(stmt)
+            rows: list[ContainerRegistryRow] = result.scalars().all()
+            return [row.to_dataclass() for row in rows]
+
+    async def mark_images_as_deleted(
+        self,
+        registry_name: str,
+        project: Optional[str] = None,
+    ) -> None:
+        """Mark all images as deleted for a given registry and optional project."""
+        async with self._db.begin_session() as session:
+            update_stmt = (
+                sa.update(ImageRow)
+                .where(ImageRow.registry == registry_name)
+                .where(ImageRow.status != ImageStatus.DELETED)
+                .values(status=ImageStatus.DELETED)
+            )
+            if project:
+                update_stmt = update_stmt.where(ImageRow.project == project)
+
+            await session.execute(update_stmt)
+
+    async def fetch_known_registries(self) -> dict[str, str]:
+        """Fetch all known container registries from the database."""
+        async with self._db.begin_readonly_session() as session:
+            known_registries_map = await ContainerRegistryRow.get_known_container_registries(
+                session
+            )
```

Comment:
Please avoid returning a dict. You should properly define and pass the data class from where you first read the value, so that subsequent logic becomes easier to read.